ethereum/stratum/src/main/java/org/hyperledger/besu/ethereum/stratum/GetWorkProtocol.java

```diff
@@ -37,50 +39,34 @@
 public class GetWorkProtocol implements StratumProtocol {
   private static final Logger LOG = LoggerFactory.getLogger(GetWorkProtocol.class);
   private static final ObjectMapper mapper = new ObjectMapper();
-  private static final String CRLF = "\r\n";
 
   private final EpochCalculator epochCalculator;
   private volatile PoWSolverInputs currentInput;
   private Function<PoWSolution, Boolean> submitCallback;
   private String[] getWorkResult;
 
-  public GetWorkProtocol(final MiningCoordinator miningCoordinator) {
-    if (miningCoordinator instanceof PoWMiningCoordinator) {
-      this.epochCalculator = ((PoWMiningCoordinator) miningCoordinator).getEpochCalculator();
-    } else {
-      this.epochCalculator = new EpochCalculator.DefaultEpochCalculator();
-    }
+  public GetWorkProtocol(final EpochCalculator epochCalculator) {
+    this.epochCalculator = epochCalculator;
   }
 
-  private JsonNode readMessage(final String message) {
-    int bodyIndex = message.indexOf(CRLF + CRLF);
-    if (bodyIndex == -1) {
-      return null;
-    }
-    if (!message.startsWith("POST / HTTP")) {
-      return null;
-    }
-    String body = message.substring(bodyIndex);
+  @Override
+  public boolean maybeHandle(
+      final Buffer initialMessage, final StratumConnection conn, final Consumer<String> sender) {
+    JsonObject message;
     try {
-      return mapper.readTree(body);
-    } catch (JsonProcessingException e) {
-      return null;
+      message = initialMessage.toJsonObject();
+    } catch (DecodeException e) {
+      return false;
     }
-  }
-
-  @Override
-  public boolean maybeHandle(final String initialMessage, final StratumConnection conn) {
-    JsonNode message = readMessage(initialMessage);
     if (message == null) {
       return false;
     }
-    JsonNode methodNode = message.get("method");
-    if (methodNode != null) {
-      String method = methodNode.textValue();
+    String method = message.getString("method");
+    if (method != null) {
       if ("eth_getWork".equals(method) || "eth_submitWork".equals(method)) {
-        JsonNode idNode = message.get("id");
-        boolean canHandle = idNode != null && idNode.isInt();
-        handle(conn, initialMessage);
+        Integer idNode = message.getInteger("id");
```

Comment:
this code is not safe. See https://vertx.io/docs/apidocs/io/vertx/core/json/JsonObject.html#getInteger-java.lang.String- - you will get a `ClassCastException` if the value is not an integer.

Reply:
I changed it now so it always tries to execute the `handle` method.